strawberry/dataloader.py

```diff
@@ -47,9 +50,31 @@ class DataLoader(Generic[K, T]):
     cache: bool = False
     cache_map: Dict[K, Future]
 
+    @overload
+    def __init__(
+        self: "DataLoader[K, T]",
+        # any BaseException is rethrown in 'load', so should be excluded from the T type
+        load_fn: Callable[[List[K]], Awaitable[Sequence[Union[T, BaseException]]]],
+        max_batch_size: Optional[int] = None,
+        cache: bool = True,
+        loop: AbstractEventLoop = None,
+    ) -> None:
+        ...
+
+    # fallback if load_fn is untyped and there's no other info for inference
+    @overload
+    def __init__(
+        self: "DataLoader[K, Any]",
+        load_fn: Callable[[List[K]], Awaitable[List[Any]]],
+        max_batch_size: Optional[int] = None,
+        cache: bool = True,
+        loop: AbstractEventLoop = None,
+    ) -> None:
+        ...
+
     def __init__(
         self,
-        load_fn: Callable[[List[K]], Awaitable[List[T]]],
+        load_fn: Callable[[List[K]], Awaitable[Sequence[Union[T, BaseException]]]],
         max_batch_size: Optional[int] = None,
         cache: bool = True,
         loop: AbstractEventLoop = None,
```

Comment:
I usually remove the types from the non overloaded method so it doesn't have to be in sync with the rest :)

Reply:
Ah. I tried what I think you're meaning (let me know if I've misinterpreted) and it looks like the parameters end up typed as `Any` in each overload. Specifically, I switched these to:

```python
    @overload
    def __init__(
        self: "DataLoader[K, T]",
        # any BaseException is rethrown in 'load', so should be excluded from the T type
        load_fn: Callable[[List[K]], Awaitable[Sequence[Union[T, BaseException]]]],
        max_batch_size = None,
        cache = True,
        loop = None,
    ) -> None:
        ...

    # fallback if load_fn is untyped and there's no other info for inference
    @overload
    def __init__(
        self: "DataLoader[K, Any]",
        load_fn: Callable[[List[K]], Awaitable[List[Any]]],
        max_batch_size = None,
        cache = True,
        loop = None,
    ) -> None:
        ...
```

`reveal_type(DataLoader.__init__)` gives (reformatted for ease of reading):

```python
Overload(
  def (
    self: strawberry.dataloader.DataLoader[K`1, T`2],
    load_fn: def (builtins.list[K`1]) -> typing.Awaitable[typing.Sequence[Union[T`2, builtins.BaseException]]], 
    max_batch_size: Any =, 
    cache: Any =, 
    loop: Any =
  ), 
  def (
    self: strawberry.dataloader.DataLoader[K`1, Any],
    load_fn: def (builtins.list[K`1]) -> typing.Awaitable[builtins.list[Any]],
    max_batch_size: Any =, 
    cache: Any =,
    loop: Any =
  )
)
```

and a call like `loader = DataLoader(load, cache="this isn't a bool")` passes type checking.

Reply:
I meant something like this: https://github.com/strawberry-graphql/strawberry/blob/c7322c0aefc009c021b29de28f1e2a2302bc75f1/strawberry/object_type.py#L152-L190

it's not a big deal anyway 😊

Reply:
Ah, I see; thanks for explaining 👍